HA is enabled for Autoscaler too (#3192)
knative/docs: docs/serving/config-ha.md

When using a leader election HA pattern, instances of controllers are already scheduled and running inside the cluster before they are required. These controller instances compete to use a shared resource, known as the leader election lock. The instance of the controller that has access to the leader election lock resource at any given time is referred to as the leader.

> HA functionality is enabled by default for all Knative Serving components except for the `autoscaler` component.

Reviewer: That should remain. By default it is still 1, no?

Author: Then I don't understand why we say HA is enabled by default. By default it is 1 replica for all components, no?

Reviewer: No i think it's 3

Author: No, I don't find any replicas or HPA in https://github.com/knative/serving/blob/master/config/core/deployments/controller.yaml. 3 is the number we set for e2e tests: https://github.com/knative/serving/blob/master/test/e2e-common.sh#L515

Reviewer: hm, I thought we did. But then sg.

Reviewer: Probably we need to rephrase. That is allowed, but not enabled by default?

Author: done

Reviewer 2: Is there any need to include this line at all now that it's available for all components? We already state above:

"Active/passive HA in Knative is available through leader election, which can be enabled after Knative Serving control plane is installed."

Also, if it's not enabled by default for everything, it now seems illogical that we have a section about disabling it immediately after this - shouldn't we have a section about enabling it first?

Author: There are two concepts here: leader election and HA. I reworded this line. Please take another look.

Reviewer 2: I think we should also remove the line above:
"Active/passive HA in Knative is available through leader election, which can be enabled after Knative Serving control plane is installed."

This made sense at the time, but doesn't anymore if we're saying leader election is enabled by default.

> Leader election is enabled but HA functionanlity is disabled with 1 replica by default for all Knative Serving components.

Reviewer 2: ```suggestion
Leader election is enabled by default for all Knative Serving components.
HA functionality is disabled by default for all Knative Serving components, which are configured with only one replica.
```
I feel like we should have a note about enabling HA though by changing the replica config or point to docs about it here?

Author: hmm, like `See Scaling the control plane section for how to enable HA`?

- Setting `--replicas=1` disables HA.
- Passing `--disable-ha` to the controller process disables leader election.

> If you scale down the `autoscaler` component, you may get inaccurate autoscaling results for a subset of your Revisions for a peroid, which can

Reviewer: ```suggestion
NOTE: if you scale down the `autoscaler` component, you may get inaccurate autoscaling results for a subset of your Revisions for a period, which can
```

> be up to `stable-window` time. This is because when an `autoscaler` pods is terminating, the leadership of its owned Revisions will be gained by

Reviewer: ```suggestion
be up to `stable-window` time. This is because when an `autoscaler` pod is terminating, the ownership of the currently owned Revisions will be acquired by
```
or `pods are`

> other standby `autoscaler` pods. The new leader `autoscaler` pods need `stable-window` time to build the scaling metrics state for its newly

Reviewer: ```suggestion
other standby `autoscaler` pods. The new leader `autoscaler` pods will need then `stable-window` time to build the scaling metrics state for those revisions.
```

- Setting `--replicas=1` disables HA.
- Passing `--disable-ha` to the controller process disables leader election.

> NOTE: if you scale down the `autoscaler` component, you may get inaccurate autoscaling results for a subset of your Revisions for a period, which can
> be up to the `stable-window` time. This is because when an `autoscaler` pod is terminating, the ownership of the currently owned Revisions will be acquired by
> other standby `autoscaler` pods. The new leader `autoscaler` pods will need the `stable-window` time to build the scaling metrics state for those Revisions.

Reviewer 2: ```suggestion
**NOTE:** If you scale down the `autoscaler` component, you may observe inaccurate autoscaling results for some revisions for a period of time up to the `stable-window` value. This is because when an `autoscaler` pod is terminating, ownership of the revisions belonging to that pod is passed to other `autoscaler` pods that are on stand by. The `autoscaler` pods that take over ownership of those revisions use the `stable-window` time to build the scaling metrics state for those revisions.
```

Reviewer 2: Tried to reword this so it makes a bit more sense. Also, use lower case where we're not referring to a specific Kubernetes object.
E.g...
The revision... vs
The `Revision` object....

- Passing `--disable-ha` to the controller process disables leader election.

NOTE: if you scale down the `autoscaler` component, you may get inaccurate autoscaling results for a subset of your Revisions for a period, which can
be up to `stable-window` time. This is because when an `autoscaler` pod is terminating, the ownership of the currently owned Revisions will be acquired by

> other standby `autoscaler` pods. The new leader `autoscaler` pods need `stable-window` time to build the scaling metrics state for its newly

Reviewer: ```suggestion
other standby `autoscaler` pods. The new leader `autoscaler` pods will need `stable-window` time to build the scaling metrics state for its newly
```
